homeassistant/components/iotawatt/__init__.py

```diff
@@ -0,0 +1,149 @@
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the iotawatt component."""
+    hass.data.setdefault(DOMAIN, {})
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up iotawatt from a config entry."""
+    polling_interval = entry.options.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
+
+    session = AsyncClient()
+    if "username" in entry.data.keys():
+        api = Iotawatt(
+            entry.data["name"],
+            entry.data["host"],
+            session,
+            entry.data["username"],
+            entry.data["password"],
+        )
+    else:
+        api = Iotawatt(
+            entry.data["name"],
+            entry.data["host"],
+            session,
+        )
+
+    coordinator = IotawattUpdater(
+        hass,
+        api=api,
+        name="IoTaWatt",
+        update_interval=polling_interval,
+    )
+
+    await coordinator.async_refresh()
+
+    if not coordinator.last_update_success:
+        raise ConfigEntryNotReady
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        COORDINATOR: coordinator,
+        IOTAWATT_API: api,
+    }
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    return True
+
+
+class IotawattUpdater(DataUpdateCoordinator):
+    """Class to manage fetching update data from the IoTaWatt Energy Device."""
+
+    def __init__(
+        self, hass: HomeAssistant, api: str, name: str, update_interval: int
+    ) -> None:
+        """Initialize IotaWattUpdater object."""
+        self.api = api
+        self.sensorlist: dict[str, list[str]] = {}
+
+        super().__init__(
+            hass=hass,
+            logger=_LOGGER,
+            name=name,
+            update_interval=timedelta(seconds=update_interval),
+        )
+
+    async def _async_update_data(self):
+        """Fetch sensors from IoTaWatt device."""
+
+        await self.api.update()
+        sensors = self.api.getSensors()
+
+        for sensor in sensors["sensors"]:
+            if sensor not in self.sensorlist:
+                to_add = {
+                    "entity": sensor,
+                    "mac_address": sensors["sensors"][sensor].hub_mac_address,
+                    "name": sensors["sensors"][sensor].getName(),
+                }
+                async_dispatcher_send(self.hass, SIGNAL_ADD_DEVICE, to_add)
+                self.sensorlist[sensor] = sensors["sensors"][sensor]
+
+        return sensors
+
+
+class IotaWattEntity(CoordinatorEntity, SensorEntity):
+    """Defines the base IoTaWatt Energy Device entity."""
+
+    def __init__(self, coordinator: IotawattUpdater, entity, mac_address, name):
+        """Initialize the IoTaWatt Entity."""
+        super().__init__(coordinator)
+
+        self._entity = entity
+        self._name = name
+        self._icon = DEFAULT_ICON
+        self._mac_address = mac_address
```

Comment:
```suggestion
        self._mac_address = mac_address
        self._attr_unique_id = mac_address
```

Reply:
I don't see how `mac_address` can be unique for every entity?

Reply:
@balloob currently the ```IotaWattEntity``` is the base class and is extended in the ```sensor.py``` file in the class ```IoTaWattSensor```. In this class the ```unique_id``` method is overwritten returning a unique ID for each sensor.

Reply:
Renamed ```self._mac_address``` to ```self._attr_unique_id```. 

Can resolve conversation if @balloob is okay with my explanation of the unique id for the base class.